Mathlib/Geometry/Manifold/IntegralCurve.lean

```diff
@@ -0,0 +1,297 @@
+    IsIntegralCurveAt γ v t := by
+  rw [Metric.mem_nhds_iff] at hs
+  obtain ⟨ε, hε, hmem⟩ := hs
+  exact ⟨ε, hε, Real.ball_eq_Ioo _ _ ▸ h.mono hmem⟩
+
+lemma IsIntegralCurveAt.isIntegralCurveOn (h : ∀ t ∈ s, IsIntegralCurveAt γ v t) :
+    IsIntegralCurveOn γ v s := by
+  intros t ht
+  obtain ⟨ε, hε, h⟩ := h t ht
+  exact h t (Real.ball_eq_Ioo _ _ ▸ Metric.mem_ball_self hε)
+
+/-! ### Translation lemmas -/
+section Translation
+
+lemma IsIntegralCurveOn.comp_add (hγ : IsIntegralCurveOn γ v s) (dt : ℝ) :
+    IsIntegralCurveOn (γ ∘ (· + dt)) v { t | t + dt ∈ s } := by
+  intros t ht
+  rw [Function.comp_apply,
+    ← ContinuousLinearMap.comp_id (ContinuousLinearMap.smulRight 1 (v (γ (t + dt))))]
+  apply HasMFDerivAt.comp t (hγ (t + dt) ht)
+  refine ⟨(continuous_add_right _).continuousAt, ?_⟩
+  simp only [mfld_simps, hasFDerivWithinAt_univ]
+  exact HasFDerivAt.add_const (hasFDerivAt_id _) _
+
+lemma isIntegralCurveOn_comp_add {dt : ℝ} :
+    IsIntegralCurveOn γ v s ↔ IsIntegralCurveOn (γ ∘ (· + dt)) v { t | t + dt ∈ s } := by
+  refine ⟨fun hγ => hγ.comp_add _, fun hγ => ?_⟩
+  have := hγ.comp_add (-dt)
+  simp only [mem_setOf_eq, neg_add_cancel_right, setOf_mem_eq] at this
+  convert this
+  ext
+  simp only [Function.comp_apply, neg_add_cancel_right]
+
+lemma IsIntegralCurveAt.comp_add (hγ : IsIntegralCurveAt γ v t₀) (dt : ℝ) :
+    IsIntegralCurveAt (γ ∘ (· + dt)) v (t₀ - dt) := by
+  obtain ⟨ε, hε, h⟩ := hγ
+  refine ⟨ε, hε, ?_⟩
+  convert h.comp_add dt
+  ext t'
+  rw [sub_right_comm, sub_add_eq_add_sub, ← add_mem_Ioo_iff_left]
+  rfl
+
+lemma isIntegralCurveAt_comp_add {dt : ℝ} :
+    IsIntegralCurveAt γ v t₀ ↔ IsIntegralCurveAt (γ ∘ (· + dt)) v (t₀ - dt) := by
+  refine ⟨fun hγ => hγ.comp_add _, fun hγ ↦ ?_⟩
+  have := hγ.comp_add (-dt)
+  rw [sub_neg_eq_add, sub_add_cancel] at this
+  convert this
+  ext
+  simp only [Function.comp_apply, neg_add_cancel_right]
+
+lemma IsIntegralCurve.comp_add (hγ : IsIntegralCurve γ v) (dt : ℝ) :
+    IsIntegralCurve (γ ∘ (· + dt)) v := by
+  rw [isIntegralCurve_iff_isIntegralCurveOn] at *
+  exact hγ.comp_add _
+
+lemma isIntegralCurve_comp_add {dt : ℝ} :
+    IsIntegralCurve γ v ↔ IsIntegralCurve (γ ∘ (· + dt)) v := by
+  refine ⟨fun hγ => hγ.comp_add _, fun hγ ↦ ?_⟩
+  convert hγ.comp_add (-dt)
+  ext
+  simp only [Function.comp_apply, neg_add_cancel_right]
+end Translation
+
+/-! ### Scaling lemmas -/
+section Scaling
+
+lemma IsIntegralCurveOn.comp_mul (hγ : IsIntegralCurveOn γ v s) (a : ℝ) :
+    IsIntegralCurveOn (γ ∘ (· * a)) (a • v) { t | t * a ∈ s } := by
+  intros t ht
+  rw [Function.comp_apply, Pi.smul_apply, ← ContinuousLinearMap.smulRight_comp]
+  refine HasMFDerivAt.comp t (hγ (t * a) ht) ⟨(continuous_mul_right _).continuousAt, ?_⟩
+  simp only [mfld_simps, hasFDerivWithinAt_univ]
+  exact HasFDerivAt.mul_const' (hasFDerivAt_id _) _
+
+lemma isIntegralCurvOn_comp_mul_ne_zero {a : ℝ} (ha : a ≠ 0) :
```

Comment:
I have the nagging feeling the next four lemmas could be golfed (all the manipulations around multiplication and division). I might be wrong; perhaps none of  simp, ring and (g)congr is useful here. It might be worth extracting an MWE and asking the hive mind on zulip.